File: leetcode/problems/go/2095. Delete the Middle Node of a Linked List/solve.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteMiddle(head *ListNode) *ListNode {
	fast := head
	slow := head
	var prev *ListNode

	if head == nil || head.Next == nil {
		return nil
	}

	for {
		fast = fast.Next
		if fast == nil {
			prev.Next = slow.Next
			break
		}
		prev = slow
		slow = slow.Next
		fast = fast.Next

		if fast == nil {
			prev.Next = slow.Next
			break
		}
	}

	return head
}

File: leetcode/problems/python3/2095. Delete the Middle Node of a Linked List/solve.py
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    # Definition for singly-linked list.
    class ListNode:
        def __init__(self, val=0, next=None):
            self.val = val
            self.next = next


class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fast = head
        slow = head
        prev = None

        if head is None or head.next is None:
            return None

        while True:
            fast = fast.next
            if fast is None:
                prev.next = slow.next
                break
            prev = slow
            slow = slow.next
            fast = fast.next
            if fast is None:
                prev.next = slow.next
                break

        return head
